core/src/main/java/org/kiwix/kiwixmobile/core/history/HistoryActivity.kt

```diff
@@ -0,0 +1,173 @@
+      }
+
+      override fun onDestroyActionMode(mode: ActionMode) {
+        historyViewModel.actions.offer(Action.ExitActionModeMenu)
+        actionMode = null
+      }
+    }
+
+  private val historyAdapter: HistoryAdapter by lazy {
+    HistoryAdapter(HistoryItemDelegate(this), HistoryDateDelegate())
+  }
+
+  override fun injection(coreComponent: CoreComponent) {
+    activityComponent.inject(this)
+  }
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_history)
+    setSupportActionBar(toolbar)
+
+    supportActionBar?.setDisplayHomeAsUpEnabled(true)
+    supportActionBar?.setTitle(string.history)
+
+    recycler_view.adapter = historyAdapter
+    recycler_view.layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)
+
+    history_switch.setOnCheckedChangeListener { _, isChecked ->
+      historyViewModel.actions.offer(ToggleShowHistoryFromAllBooks(isChecked))
+    }
+    history_switch.isChecked = !sharedPreferenceUtil.showHistoryCurrentBook
+
+    compositeDisposable.add(historyViewModel.effects.subscribe { it.invokeWith(this) })
+  }
+
+  override fun onDestroy() {
+    compositeDisposable.clear()
+    super.onDestroy()
+  }
+
+  override fun onCreateOptionsMenu(menu: Menu): Boolean {
+    menuInflater.inflate(R.menu.menu_history, menu)
+    val search = menu.findItem(R.id.menu_history_search).actionView as SearchView
+    search.queryHint = getString(R.string.search_history)
+    search.setOnQueryTextListener(SimpleTextListener {
+      historyViewModel.actions.offer(Filter(it))
+    })
+    historyViewModel.state.observe(this, Observer(::render))
+    return true
+  }
+
+  override fun onOptionsItemSelected(item: MenuItem): Boolean {
+    if (item.itemId == android.R.id.home) {
+      historyViewModel.actions.offer(ExitHistory)
+    }
+    if (item.itemId == R.id.menu_history_clear) {
+      historyViewModel.actions.offer(RequestDeleteAllHistoryItems(dialogShower))
+    }
+    return super.onOptionsItemSelected(item)
+  }
+
+  private fun render(state: State) =
+    when (state) {
+      is Results -> {
+        actionMode?.finish()
+        historyAdapter.items = state.historyItems
+        history_switch.isEnabled = true
+        no_history.visibility = View.GONE
+      }
+      is SelectionResults -> {
+        if (state.selectedHistoryItems.isNotEmpty() && actionMode == null) {
+          actionMode = startSupportActionMode(actionModeCallback)
+        }
+        historyAdapter.items = state.historyItems
+        history_switch.isEnabled = false
+        no_history.visibility = View.GONE
+      }
+      is NoResults -> {
+        historyAdapter.items = state.historyItems
+        no_history.visibility = View.VISIBLE
+      }
+    }
+
+  override fun onItemClick(
+    favicon: ImageView,
+    history: HistoryItem
+  ) {
+    historyViewModel.actions.offer(Action.OnItemClick(history))
+  }
+
+  override fun onItemLongClick(
```

Comment:
`ctrl+shift+j` a few timesto get these all on the same line, looks like they'll fit